Change default IO type from NETCDF4C to PNETCDF

amametjanov opened the pull request:
Change default IO type from NETCDF4C to PNETCDF

<!--
Below are a few things we ask you or your reviewers to kindly check.
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Linting
  * [x] [Pre-commit](https://docs.e3sm.org/Omega/Omega/devGuide/Linting.html) run locally
* [x] Building
  * [x] CMake build does not produce any new warnings from changes in this PR
* [ ] Testing
  * [ ] Add a comment to the PR titled `Testing` with the following:
    * [x] Which machines [CTest unit tests](https://docs.e3sm.org/Omega/Omega/devGuide/QuickStart.html#running-ctests)
          have been run on and indicate that are all passing.
    * [ ] The [Polaris omega_pr test suite](https://docs.e3sm.org/Omega/Omega/devGuide/Testing.html)
       has passed, using the Polaris `e3sm_submodules/Omega` baseline
    * [ ] Document machine(s), compiler(s), and the build path(s) used for `-p` for both the baseline (Polaris `e3sm_submodules/Omega`) and the PR build
    * [x] Indicate "All tests passed" or document failing tests
    * [ ] Document testing used to verify the changes including any tests that are added/modified/impacted.
    * [ ] Performance related PRs: Please include a relevant PACE experiment link documenting performance before and after.

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

Fixes #333 
Closes #334

amametjanov committed:
Change default IO type from NETCDF4C to PNETCDF

amametjanov commented:
Testing:
- aurora+oneapi-ifxgpu: `100% tests passed, 0 tests failed out of 35`
- frontier:
  * craygnu: `100% tests passed`
  * craygnu-mphipcc: `100%`
  * craycray: `80% tests passed, 7 tests failed out of 35` (unrelated to IO)
  * craycray-mphipcc: `100%`
  * crayamd: `100%`
  * crayamd-mphipcc: `100%`
- pm:
  * gnu: `100%`
  * intel: `100%`
  * gnugpu: `100%`
- chrysalis:
  * intel: `100%`
  * gnu: `100%`

xylar commented:
Thanks very much, @amametjanov!  This looks promising.

xylar commented:
@amametjanov, I've got 3 tests for this in the queue, one on Chrysalis and 2 on Frontier.  But wait times seem to be a bit long both places.  I'll keep you posted.

amametjanov commented:
To fix `IO_Test` ctest, this PR needs [scorpio PR #670](https://github.com/E3SM-Project/scorpio/pull/670), which expands supports for CDF5 data-types (like 64-bit ints).

xylar commented:
@amametjanov, maybe this will be fixed by https://github.com/E3SM-Project/scorpio/pull/670 but what I'm seeing on Chrysalis with this branch is:
```
PIO: FATAL ERROR: Aborting... An error occured, Waiting on pending requests on file (output.nc, ncid=21) failed (Number of pending requests on file = 1, Number of variables with pending requests = 1, Number of request blocks = 1, Current block being waited on = 0, Number of requests in current block = 1).. NetCDF: Operation not allowed in define mode (err=-39). Aborting since the error handler was set to PIO_INTERNAL_ERROR... (/home/ac.xylar/e3sm_work/e3sm/azamat/mod-dflt-io-type/externals/scorpio/src/clib/pio_darray_int.cpp: 2192)
```
The polaris output is available at:
```
/lcrc/group/e3sm/ac.xylar/polaris_0.9/chrysalis/test_20251209/omega-pr-mod-dflt-io-type
```
It seems like this won't be a short-term fix for Omega if a scoprio fix is needed, because that would mean:
* scorpio fix gets merge
* scorpio release happens
* scorpio gets update in E3SM
* E3SM/master gets merged into Omega/develop
* This branch goes in

It feels like we should look into whether there's some alternative way to address #323 in the next week or two.

amametjanov committed:
Update scorpio

Update scorpio from v1.8.2 2025-Jul-14 to v1.9.0 2025-Nov-21.
Also add fix for PnetCDF CDF5 types.

amametjanov committed:
Add end-define-mode calls while in data-mode netcdf writes

amametjanov commented:
Xylar, please check with updated head of this branch to see if it fixes `NetCDF: Operation not allowed in define mode` errors in polaris tests. All ctests are passing for me.

This branch updates the scorpio submodule ahead of E3SM/master's version, which is still on v1.8.2 2025-Jul-14. When scorpio gets update in E3SM/master (with v1.9.0 or later), E3SM/master merge to Omega/develop will subsume this branch's updates.

xylar commented:
Thanks, @amametjanov!  I'll retest as soon as I can.

xylar commented on components/omega/configs/Default.yml:
```diff
@@ -129,7 +129,7 @@ Omega:
       Filename: ocn.hifreq.$Y-$M
       Mode: write
-      IfExists: append
+      IfExists: replace
```
I don't feel qualified to decide if this is a change we want or not.  It seems potentially outside the scope of this particular PR even if we do.

xylar commented on components/omega/src/base/IO.cpp:
```diff
@@ -269,7 +269,7 @@ void openFile(
       case IfExists::Fail:
          PIOErr = PIOc_createfile(SysID, &FileID, &Format, Filename.c_str(),
-                                  NC_NOCLOBBER | InMode);
+                                  NC_NOCLOBBER | PIO_64BIT_DATA | InMode);
```
Here and elsewhere, what are the implications of this change if someone deliberately chooses a different IO format than the new default at runtime?  It seems like something we might not want to be hard-coding in this way but maybe I don't understand the implications.

xylar reviewed (commented):
I tested the `omega_pr` suite using the fix in https://github.com/E3SM-Project/polaris/pull/442, pointing to this branch for the Omega build.

I was able to run successfully with both Intel and Gnu on Chrysalis.  I discovered that I can't log in to either Aurora or Frontier at the moment.  I'm trying on Perlmutter (CPU and GPU) next.

In the mean time, two small questions/comments.

xylar commented:
On Perlmutter-CPU (Both Intel and Gnu) and -GPU (Gnu-GPU), I'm seeing the same hanging behavior reported in https://github.com/E3SM-Project/polaris/issues/396 as we had seen previously.  It seems like maybe that behavior is unfortunately independent of this PIO problem.

grnydawn commented:
@amametjanov , I ran the tests for this PR on Frontier, but I got the same PIO error: 
```
PIO: FATAL ERROR: Aborting... invalid IO type (/lustre/orion/cli115/scratch/grnydawn/cronjobs/tasks/polaris_cdash/polaris/e3sm_submodules/Omega/externals/scorpio/src/clib/pio_lists.cpp: 120)
```
Please see Frontier test results at Omega CDash dashboard at https://my.cdash.org/index.php?project=omega

amametjanov committed:
Remove colon from ocn.hist.$SimTime for MPI-IO on Frontier

Also, address GH-review comments.

grnydawn committed:
fix a bug in reading IODefaultFormat from omega.yml

amametjanov committed:
Update netcdf4->pnetcdf in default yaml and if no io-yaml exists

xylar commented:
@amametjanov, could you please let me know when this is ready for me to re-test in Polaris?

amametjanov commented:
Yes, this is ready, please re-run Polaris tests. :pray:

xylar commented:
## Testing

I successfully ran the `omega_pr` suite from Polaris using this branch and:
- [x] Chrysalis with intel
- [x] Chrysalis with gnu
- [x] Frontier with craygnu
- [ ] Frontier with craygnu-mphipcc
- [x] Frontier with craycray
- [x] Frontier with craycray-mphipcc
- [x] Frontier with crayamd
- [ ] Frontier with crayamd-mphipcc

(Unchecked items are still in the queue -- update soon...)

I also verified that one `output.nc` file produced by Omega was in `cdf5` format, as expected.  I presume the same will be true for all files.

@amametjanov and @grnydawn, thank you so much for figuring out these issues and fixing them!

xylar commented:
My Frontier tests are now running.  All the CPU tests did okay.  The GPU tests are *very* slow by comparison, and they are taking more than the 1 hour I had allocated.  I don't know for sure but I presume the slowness is not from this PR.

I also saw the I/O failure in one `craygnu-mphipcc` test: `ocean/planar/manufactured_solution/convergence_both/del2`.  It is discouraging that I am seeing this failure but the situation is clearly far better than it was before.

I will rerun both the tests that timed out and the one that failed.  We will see what happens.

**Update**: It seems like the file system on Frontier might be a problem.  My resubmitted jobs are hanging just trying to load the environment.

xylar commented:
In the `crayamd-mphipcc` run on Frontier, I'm seeing a different error after trying to rerun following a timeout:
```
PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: Operation not allowed in data mode (file = output.nc) (/ccs/home/xylar/e3sm_work/Omega/azamat/mod-dflt-io-type/externals/scorpio/src/clib/pio_getput_int.cpp: 544)
```
This is for the `ocean/spherical/icos/rotation_2d` test case.

xylar commented:
When I try to rerun the failed  test (` ocean/planar/manufactured_solution/convergence_both/del2`) for `craygnu-mphipcc`, I see a different error:
```
PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: Unknown file format (file = output.nc) (/ccs/home/xylar/e3sm_work/Omega/azamat/mod-dflt-io-type/externals/scorpio/src/clib/pioc_support.cpp: 5503)
```

The original error was the same as we have seen before:
```
PIO: FATAL ERROR: Aborting... invalid IO type (/ccs/home/xylar/e3sm_work/Omega/azamat/mod-dflt-io-type/externals/scorpio/src/clib/pio_lists.cpp: 120)
```

I presume these errors might indicate that Omega isn't overwriting the `output.nc` file as expected if a run either times out or fails.  If so, that issue would presumably be beyond the scope of this PR.

rljacob commented:
For Frontier, craygnu and craygnu-mphipcc is the only compiler E3SM cares about.    Don't spend more then a token amount of E3SM time looking that the others.

xylar commented:
Okay, thanks @rljacob.  That wasn't clear to me.

xylar commented:
I set up `ocean/planar/manufactured_solution/convergence_both/del2` and reran it with `craygnu-mphipcc`, this time successfully, so the problem there seems to be transient.

xylar commented:
Nope, now I'm seeing the usual error message:
```
PIO: FATAL ERROR: Aborting... invalid IO type (/ccs/home/xylar/e3sm_work/Omega/azamat/mod-dflt-io-type/externals/scorpio/src/clib/pio_lists.cpp: 120)
```
but this time in the `ocean/planar/manufactured_solution/convergence_both/default` (again with `craygnu-mphipcc`).

So I think this PR should go in but we can't consider this problem to be solved.

amametjanov commented:
Thank you for re-running Polaris tests (and merging).
I re-ran omega ctests on latest post-merge `develop` with all 6 frontier compilers and saw just 2 ctests fail with `craycray` on cpus; all others pass:
- `11 - HORZOPERATORS_PLANE_TEST` and `12 - HORZOPERATORS_SPHERE_TEST`

I heard that frontier scratch filesystem was hanging and slow this week: maybe that's the culprit.
